front-end-tools/src/components/DeployComponent.tsx

```diff
@@ -0,0 +1,268 @@
+                    })
+                    .catch((e) => {
+                        setModuleReferenceDeployed(undefined);
+                        setViewErrorModuleReference((e as Error).message);
+                        clearInterval(interval);
+                    });
+            }, REFRESH_INTERVAL.asMilliseconds());
+        }
+    }, [connection, account, client, txHashDeploy]);
+
+    useEffect(() => {
+        if (connection && client && account && moduleReferenceCalculated) {
+            client
+                .getModuleSource(new ModuleReference(moduleReferenceCalculated))
+                .then((value) => {
+                    if (value === undefined) {
+                        setIsModuleReferenceAlreadyDeployedStep1(false);
+                    } else {
+                        setIsModuleReferenceAlreadyDeployedStep1(true);
+                    }
+                })
+                .catch(() => {
+                    setIsModuleReferenceAlreadyDeployedStep1(false);
+                });
+        }
+    }, [connection, account, client, moduleReferenceCalculated]);
+
+    function onSubmit() {
+        setTxHashDeploy(undefined);
+        setTransactionErrorDeploy(undefined);
+
+        // Send deploy transaction
+
+        const tx = deploy(connection, account, base64Module);
+        tx.then((txHash) => {
+            setModuleReferenceDeployed(undefined);
+            setTxHashDeploy(txHash);
+        }).catch((err: Error) => setTransactionErrorDeploy((err as Error).message));
+    }
+
+    return (
+        <Box header="Step 1: Deploy Smart Contract Module">
+            <Form onSubmit={form.handleSubmit(onSubmit)}>
+                <Form.Group className="mb-3">
+                    <Form.Label>Upload Smart Contract Module File (e.g. myContract.wasm.v1)</Form.Label>
+                    <Form.Control
+                        type="file"
+                        accept=".wasm,.wasm.v0,.wasm.v1"
+                        {...form.register('file')}
+                        onChange={async (e) => {
+                            const register = form.register('file');
+
+                            register.onChange(e);
+
+                            setUploadError(undefined);
+                            setModuleReferenceDeployed(undefined);
+                            setTransactionErrorDeploy(undefined);
+                            setTxHashDeploy(undefined);
+
+                            const files = form.getValues('file');
+
+                            if (files !== undefined && files !== null && files.length > 0) {
+                                const file = files[0];
+                                const arrayBuffer = await file.arrayBuffer();
+
+                                const module = btoa(
+                                    new Uint8Array(arrayBuffer).reduce((data, byte) => {
+                                        return data + String.fromCharCode(byte);
+                                    }, '')
```

Comment:
I don't understand why we use `fromCharCode` here. If that's **the way** that's fine, but please add a comment since it's an odd construction.

Reply:
I recall that initially, I was converting the whole file in one go but that didn't work for our largest smart contracts. The `reduce` construct enables the conversion in smaller junks and the error went away.